File: packages/dcflight/android/bin/src/main/kotlin/com/dotcorr/dcflight/context/DCFContext.kt
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.dotcorr.dcflight.context

import android.content.Context
import com.dotcorr.dcflight.theme.DCFTheme

/**
 * Context interface for DCFlight components
 * Provides access to Android context, theme, and app-level state
 * Similar to React Native's ReactContext hierarchy
 */
interface DCFContext {
    val context: Context
    val isDarkMode: Boolean
    val textColor: Int
    val backgroundColor: Int
    val surfaceColor: Int
    val accentColor: Int
}

/**
 * Application-level context (root context)
 * Provides access to Android application context and theme
 */
class DCFApplicationContext(
    override val context: Context
) : DCFContext {
    override val isDarkMode: Boolean
        get() = DCFTheme.isDarkTheme(context)
    
    override val textColor: Int
        get() = DCFTheme.getTextColor(context)
    
    override val backgroundColor: Int
        get() = DCFTheme.getBackgroundColor(context)
    
    override val surfaceColor: Int
        get() = DCFTheme.getSurfaceColor(context)
    
    override val accentColor: Int
        get() = DCFTheme.getAccentColor(context)
}

/**
 * Themed context for specific surfaces/views
 * Can override theme values for specific contexts
 */
class DCFThemedContext(
    private val parent: DCFContext,
    override val isDarkMode: Boolean = parent.isDarkMode,
    override val textColor: Int = parent.textColor,
    override val backgroundColor: Int = parent.backgroundColor,
    override val surfaceColor: Int = parent.surfaceColor,
    override val accentColor: Int = parent.accentColor
) : DCFContext {
    override val context: Context
        get() = parent.context
}

File: packages/dcflight/ios/Classes/Context/DCFContext.swift
/*
 * Copyright (c) Dotcorr Studio. and affiliates.
 *
 * Licensed under the PolyForm Noncommercial License 1.0.0.
 * Commercial use requires a license from DotCorr.
 */

import UIKit

/// Context protocol for DCFlight components
/// Provides access to view controller, theme, and app-level state
/// Context hierarchy for managing component context and state
public protocol DCFContext {
    var viewController: UIViewController? { get }
    var isDarkMode: Bool { get }
    var textColor: UIColor { get }
    var backgroundColor: UIColor { get }
    var surfaceColor: UIColor { get }
    var accentColor: UIColor { get }
}

/// Application-level context (root context)
/// Provides access to app-level context and theme
public class DCFApplicationContext: DCFContext {
    public let viewController: UIViewController? = nil
    
    private let traitCollection: UITraitCollection
    
    public init(traitCollection: UITraitCollection = UITraitCollection.current) {
        self.traitCollection = traitCollection
    }
    
    public var isDarkMode: Bool {
        return DCFTheme.isDarkTheme(traitCollection: traitCollection)
    }
    
    public var textColor: UIColor {
        return DCFTheme.getTextColor(traitCollection: traitCollection)
    }
    
    public var backgroundColor: UIColor {
        return DCFTheme.getBackgroundColor(traitCollection: traitCollection)
    }
    
    public var surfaceColor: UIColor {
        return DCFTheme.getSurfaceColor(traitCollection: traitCollection)
    }
    
    public var accentColor: UIColor {
        return DCFTheme.getAccentColor(traitCollection: traitCollection)
    }
}

/// Themed context for specific surfaces/views
/// Can override theme values for specific contexts
public class DCFThemedContext: DCFContext {
    public let viewController: UIViewController?
    
    private let parent: DCFContext
    
    public let isDarkMode: Bool
    public let textColor: UIColor
    public let backgroundColor: UIColor
    public let surfaceColor: UIColor
    public let accentColor: UIColor
    
    public init(
        parent: DCFContext,
        viewController: UIViewController? = nil,
        isDarkMode: Bool? = nil,
        textColor: UIColor? = nil,
        backgroundColor: UIColor? = nil,
        surfaceColor: UIColor? = nil,
        accentColor: UIColor? = nil
    ) {
        self.parent = parent
        self.viewController = viewController
        self.isDarkMode = isDarkMode ?? parent.isDarkMode
        self.textColor = textColor ?? parent.textColor
        self.backgroundColor = backgroundColor ?? parent.backgroundColor
        self.surfaceColor = surfaceColor ?? parent.surfaceColor
        self.accentColor = accentColor ?? parent.accentColor
    }
}

